### PR TITLE
Feature AppEvents

### PR DESCRIPTION
**Pull request by oodavid**
I've exposed the AppEvents methods and pre-defined constants for Android and iOS. Not yet tested it in a Facebook Canvas environment, I will look into that in due course.

https://github.com/gameclosure/devkit/issues/221

### CODE REVIEW

File: ios/FacebookPlugin.mm [hunk 1]
```diff
@@ -282,6 +282,40 @@ - (void) getAuthResponse:(NSDictionary *)opts withRequestId:(NSNumber *)requestI
   [[PluginManager get] dispatchJSResponse:authResponse withError:nil andRequestId:requestId];
 }
 
+- (void) logEvent:(NSDictionary *)opts {
+  NSLOG(@"{facebook} logEvent");
+  NSString     * eventName  = [opts objectForKey:@"eventName"];
+  NSNumber     * valueToSum = [opts objectForKey:@"valueToSum"];
+  NSDictionary * parameters = [opts objectForKey:@"parameters"];
```
Comment (jwilm) [inline ios/FacebookPlugin.mm]: These should be formatted like `opts[@"key"]`

File: ios/FacebookPlugin.mm [hunk 2]
```diff
@@ -282,6 +282,40 @@ - (void) getAuthResponse:(NSDictionary *)opts withRequestId:(NSNumber *)requestI
...
+      [FBAppEvents logEvent:eventName ];
+    }
+  }
+}
+
+- (void) logPurchase:(NSDictionary *)opts {
+  NSLOG(@"{facebook} logPurchase");
+  NSNumber     *purchaseAmount = [opts objectForKey:@"purchaseAmount"];
+  NSString     *currency       = [opts objectForKey:@"currency"];
+  NSDictionary *parameters     = [opts objectForKey:@"parameters"];
```
Comment (jwilm) [inline ios/FacebookPlugin.mm]: ditto

File: ios/FacebookPlugin.mm [hunk 3]
```diff
@@ -282,6 +282,40 @@ - (void) getAuthResponse:(NSDictionary *)opts withRequestId:(NSNumber *)requestI
...
+  // Log the purchase
+  if(parameters != [NSNull null]){
+    [FBAppEvents logPurchase:[purchaseAmount doubleValue] currency:currency];
+  } else {
+    [FBAppEvents logPurchase:[purchaseAmount doubleValue] currency:currency parameters:parameters];
```
Comment (jwilm) [inline ios/FacebookPlugin.mm]: What about just setting parameters to `nil` if the test passes and then we only need the `[FBAppEvents logPurchase:[purchaseAmount doubleValue] currency:currency parameters:parameters];` call. The shorter version just delegates to this method anyway [link](https://github.com/facebook/facebook-ios-sdk/blob/master/FBSDKCoreKit/FBSDKCoreKit/FBSDKAppEvents.m#L296-L302)

File: js/native/pluginImpl.js [hunk 4]
```diff
@@ -254,6 +254,79 @@ function createNativeFacebookWrapper () {
...
+       * @param {string} eventName
+       * @param {number} valueToSum   (optional)
+       * @param {object} [parameters] (optional)
+       */
+
+      logEvent: function FBNativeLogEvent () {
+        nativeFB.notify('logEvent', {
+          "eventName":  arguments[0],
+          "valueToSum": arguments[1]||null,
+          "parameters": arguments[2]||{}
```
Comment (jwilm) [inline js/native/pluginImpl.js]: Don't need quotes here

File: js/native/pluginImpl.js [hunk 5]
```diff
@@ -254,6 +254,79 @@ function createNativeFacebookWrapper () {
...
+
+      /**
+       * @method logPurchase
+       */
+
+      logPurchase: function FBNativeLogPurchase () {
+        nativeFB.notify('logPurchase', {
+          "purchaseAmount": arguments[0],
+          "currency":       arguments[1]||"USD",
+          "parameters":     arguments[2]||{}
```
Comment (jwilm) [inline js/native/pluginImpl.js]: ditto

File: build/index.js [hunk 6]
```diff
@@ -18,7 +18,7 @@ exports.onBeforeBuild = function (api, app, config, cb) {
              'var js, fjs = d.getElementsByTagName(s)[0];',
              'if (d.getElementById(id)) {return;}',
              'js = d.createElement(s); js.id = id;',
-             'js.src = "//connect.facebook.net/en_US/all.js";',
+             'js.src = "//connect.facebook.net/en_US/sdk.js";',
```
Comment (jwilm) [inline build/index.js]: Does this still work in the simulator? I seem to recall some issues when not using `all`.

File: android/FacebookPlugin.java [hunk 7]
```diff
@@ -444,6 +450,90 @@ public void logout(String json, Integer requestId) {
...
+        log("logPurchase - error converting JSONObject to Bundle");
+      }
+    }
+
+    // Log the purchase
+    if(parameters != null){
+      aeLogger.logPurchase(BigDecimal.valueOf(purchaseAmount), Currency.getInstance(currency), parameters);
+    } else {
+      aeLogger.logPurchase(BigDecimal.valueOf(purchaseAmount), Currency.getInstance(currency));
+    }
```
Comment (jwilm) [inline android/FacebookPlugin.java]: You can just pass parameters in whether or not it's null which means this check it not needed at all on android. [link](https://github.com/facebook/facebook-android-sdk/blob/master/facebook/src/com/facebook/appevents/AppEventsLogger.java#L544-L546).

File: android/FacebookPlugin.java [hunk 8]
```diff
@@ -444,6 +450,90 @@ public void logout(String json, Integer requestId) {
...
+    JSONObject json        = null;
+    String     eventName   = null;
+    Double     valueToSum  = null;
+    JSONObject _parameters = null;
+    Bundle     parameters  = null;
+
+    // Parse JSON;
+    try {
+      json        = new JSONObject(s_json);
+      eventName   = json.optString("eventName");
```
Comment (jwilm) [inline android/FacebookPlugin.java]: Should this actually be optional?

**Comment by jwilm** [comment]
Hi @oodavid! Thanks for the PR :smile:. I commented on a few nits, and I also left a couple of questions. It looks good over all. Thanks for putting this together!

File: build/index.js (hunk 6 above)
Comment (oodavid) [inline build/index.js]: I'll run some tests ASAP, am keen to switch over to to sdk.js as all.js was deprecated last month - I think the facebook servers are routing old v1 endpoints to v2 so they probably act very similarly. That said, all.js doesn't expose AppEvents which is the driving force behind the change.

I've addressed the other snags in my next commit, but will make another pull request once I've had the chance to see what the knock-on effect of this change is.

Comment (jwilm) [inline build/index.js]: That is a fine reason to switch to sdk.js. Thanks for explaining!

**Pull request closed without merging**